### PR TITLE
http compression level 

### LINKED ISSUE #1700: Allow tuning for gzip/deflate compression level (for performance)
Allowing Vert.Xers to cofigure their webserver deflater compression level.

Inspired by the nginx equivalent: http://nginx.org/en/docs/http/ngx_http_gzip_module.html#gzip_comp_level

Usage example:

``` java
HttpServerOptions()
                .setCompressionSupported(true)
                .setCompressionLevel(1) /* valid range 1-9 (1: fastest, less compression - 9 slower, best (pedantic) compression )*/
```

**Note:**
Before this PR the .setCompressionSupported(true) was enabling the netty's HttpCompressor default compression level that is 6.

In my proposal i  set the **default compression level to 1** (fastest, not the best size saver but anyway very good compression the modern web usage case (in my experience), less cpu cylce wasting, less http requests stucking in 'waiting state')

**Why this PR?**
1. Making vert.x flexible in this configuration too cause in order to demonstrate we have nothing to envy to other webservers like nginx ;)
2. Saving server resource usage (less cpu time for compression = more cpu time for the event loop, less overwarming ;) )
3. Reducing latency TTFB for static assets when compression is enabled

**A bit more  rationale:**

A best practice suggested by any moder site 'speed test' is to enable compression, but - as a drawback - enabling it we consume more cpu cycles at server end, and if we tune up it uncarefully we can end up - paradoxically - to get a worst time-to-first-Byte (TTFB) and even a worst total request time with compression enabled compared to no compression at all.

A best pratice to avoid such issue and achieve the better performance compromise, is to tune  compressor to do the 'bare minimal' work and don't overkillig, in other words, remembering that our target is **reducing latency** at the very minimum level and not really reducing 'file size'.

From my personal experience with nginx and apache too, and confirmed by many online guides, the best value for gzip compression level is 1 or 2 for a web server. Going higher that than just saves some bits, but pay back it hardly in server resource usages and processing latency (when compression is not cached but done on-the-fly) 

Bonus: a further performance improvement can be to implement a compression cache layer in vertx but this is uncommon and complex also in other webservers so not a priority.

See also discussion on closed PR: https://github.com/eclipse/vert.x/pull/1576 (closed due to CLA and master branch  git commit history sync issues)

### PR DESCRIPTION
**Pull request by creativeprogramming**
see also: #1576
closes #1700

### CODE REVIEW

**Comment by creativeprogramming** [comment]
Except having committed my workstation hardware information for mistake (I'm a lazy git noob it's clear) now this commit should be fine.

We've both Http and Http2 Working with JUnit tests done for compression (compress/decompress) and byte size (compression_level_1 size > compression_level_9 size > 0)
ì

**Comment by creativeprogramming** [comment]
PS, OffTopic @vietj  now I understood that for 'missing client support' in #1576 discussion you probably meant something related to https://github.com/eclipse/vert.x/pull/1648/files in order to allow clients to use the same deflater in the request body. 

And I agree that it can be useful, but yes I think we can do that in a separate PR in the future, anyway as I found that HttpClients lacks of other helpers - that in the server are present instead - for instance some api to add - with ease - http post parameters to a client request body so i think for now who need that feature can wirte it's own client body handler linke in the #1648 tests  etc..

File: src/main/java/io/vertx/core/http/HttpServerOptions.java [hunk 1]
```diff
@@ -344,6 +352,24 @@ public HttpServerOptions setCompressionSupported(boolean compressionSupported) {
...
 
   /**
+   * @return the server gzip compression level
+   */
+  public int getCompressionLevel() {
+    return this.compressionLevel;
+  }
+  
+  /**
+   * Set whether the server supports compression
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/HttpServerOptions.java]: this is wrong, this set the compression level, compression is set by HttpServerOptions#setCompressionSupported(true)

**Comment by vietj** [comment]
Overall it looks good, but it is missing a couple of lines in the general documentation.

**Comment by vietj** [comment]
there is currently an effort to improve the HttpClient usability : https://groups.google.com/d/msg/vertx-dev/HN0tJs-BZqk/F2u7i4C6DQAJ

you may want to join the list and be involved!

**Comment by creativeprogramming** [comment]
@vietj I did another commit to sync with master

Let me know if you'll merge or need some change. Thanks

**Comment by vietj** [comment]
@creativeprogramming still need a couple of line in the documentation about compression level

**Comment by creativeprogramming** [comment]
@vietj i don't know if i got where you want doc improvements, I already fixed your previous comment, now I expanded the description of the most used API methods, tell me if you need more and/or feel free to change it after a merge of this PR in master

**Comment by vietj** [comment]
looks good but I actually meant at this place : https://github.com/eclipse/vert.x/blob/master/src/main/java/io/vertx/core/http/package-info.java#L1176 (I should have been more clear)

**Comment by creativeprogramming** [comment]
Done. 

PS: tell me if you really agree to have '1' as default value, it was '6' before, I want to put to '1' to automatically give benefits to upgrading users that where using compression, but anyway this is a 'sort of breaking change' there may be cases where lower size and network traffic reduced is priority so maybe it should be put in the changelog or maybe I've to set 6 as default value, final decision up to you, let me know.

**Comment by vietj** [comment]
I would rather keep the previous value as you said.

**Comment by creativeprogramming** [comment]
Ok set to 6 as per Netty legacy @ https://netty.io/4.0/api/io/netty/handler/codec/http/HttpContentCompressor.html for http/1.x (and @ https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/CompressorHttp2ConnectionEncoder.java#L43 for http/2)

**Comment by vietj** [comment]
I've seen that you use 4 tab spaces, can you reformat your PR to follow the https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines ?

**Comment by creativeprogramming** [comment]
@vietj I'm using the .editorconfig file as suggested and It reformatted me a lot of lines (effectively wrong styled in the means of https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines ) unrelated to my changes and this PR in all the involved source files.

Vert.x master is full of wrong formatted lines, I think you should do a PR for the whole vert.x for that issue and - if you find them useful - merge my changes of this PR "as they are" as that requirements where not enforced to others and we can do a separate work on that (on the whole project ) just after merging this, e.g. we can do a global reformatting of all the wrong things while adding an .editorconfig file ( ref. #1707 )

**Comment by creativeprogramming** [comment]
Just to be more clear i did a PR #1708 for showing you what I mean for "unrelated to my changes".

Please check  all the changes  the .editorconfig formatting plugin of my IDE applied to the files involved in this PR

Most is simply javadoc line truncation, but it's not only that: e.g in HttpServerImpl you'll find some code wrongly formatted for instance some `if` (blocks?) without  brackets `{}`, and without the proper spacing rules

**Comment by creativeprogramming** [comment]
To recap: You can merge this one or this: https://github.com/eclipse/vert.x/pull/1708 (formatted as you required).

**Comment by vietj** [comment]
I'm merging this one and will make some edits to save time - thanks for the good contribution @creativeprogramming

**Pull request merged**